File: SwrveSDK/SDK/Messaging/SwrveTriggerCondition.m
#import "SwrveTriggerCondition.h"

@implementation SwrveTriggerCondition

@synthesize key = _key;
@synthesize value = _value;
@synthesize triggerOperator = _triggerOperator;
@synthesize conditionOperator = _conditionOperator;

- (id) initWithDictionary:(NSDictionary *)dictionary andOperator:(NSString *) operatorKey {
    self = [super init];
    if(self) {
        _key = [dictionary objectForKey:@"key"];
        _value = [dictionary objectForKey:@"value"];
        _triggerOperator = [self determineSwrveOperator:operatorKey];
        _conditionOperator = [self determineSwrveOperator:[dictionary objectForKey:@"op"]];
        if(_key && _value && _conditionOperator){
            return self;
        }
    }
    return nil;
}

- (SwrveTriggerOperator) determineSwrveOperator:(NSString *)op {

    if([op isEqualToString:@"and"]){
        return SwrveTriggerOperatorAND;
    }else if([op isEqualToString:@"eq"]){
        return SwrveTriggerOperatorEQUALS;
    }else if([op isEqualToString:@"or"]){
        return SwrveTriggerOperatorOR;
    } else if([op isEqualToString:@"contains"]){
        return SwrveTriggerOperatorCONTAINS;
    } else if([op isEqualToString:@"number_eq"]){
        return SwrveTriggerOperatorNUMBER_EQUALS;
    } else if([op isEqualToString:@"number_gt"]){
        return SwrveTriggerOperatorNUMBER_GT;
    } else if([op isEqualToString:@"number_lt"]){
        return SwrveTriggerOperatorNUMBER_LT;
    } else if([op isEqualToString:@"number_not_between"]){
        return SwrveTriggerOperatorNUMBER_NOT_BETWEEN;
    } else if([op isEqualToString:@"number_between"]){
        return SwrveTriggerOperatorNUMBER_BETWEEN;
    } else{
        return SwrveTriggerOperatorOTHER;
    }
}

- (BOOL) hasFulfilledCondition:(NSDictionary *)payload {
    
    if (!payload) {
        return NO;
    }
    
    NSArray *payloadKeys = [payload allKeys];
    if ([payloadKeys containsObject:_key]) {
        if ([payload objectForKey:_key] != [NSNull null]) {
            id payloadObject = [payload objectForKey:_key];
            NSString *payloadValue = nil;
            if ([payloadObject isKindOfClass:[NSString class]]) {
                payloadValue = payloadObject;
            } else {
                if ([payloadObject respondsToSelector:@selector(stringValue)]) {
                    payloadValue = [(id)payloadObject stringValue];
                } else {
                    return NO;
                }
            }
            
            NSDictionary *valueDictionary = nil;
            NSString *valueString = nil;
            NSNumber *valueInteger = nil;
            if([_value isKindOfClass:[NSString class]]) {
                valueString = (NSString*)_value;
            } else if([_value isKindOfClass:[NSDictionary class]]) {
                valueDictionary = (NSDictionary*)_value;
            } else if([_value isKindOfClass:[NSNumber class]]) {
                valueInteger = (NSNumber*)_value;
            }

            if(_conditionOperator == SwrveTriggerOperatorCONTAINS) {
                return (payloadValue && [payloadValue localizedCaseInsensitiveContainsString:valueString]);
            } else if (_conditionOperator == SwrveTriggerOperatorEQUALS) {
                return (payloadValue && [payloadValue caseInsensitiveCompare:valueString] == NSOrderedSame);
            } else if (_conditionOperator == SwrveTriggerOperatorNUMBER_GT) {
                return (payloadValue && payloadValue.integerValue > valueInteger.integerValue);
            } else if (_conditionOperator == SwrveTriggerOperatorNUMBER_LT) {
                return (payloadValue && payloadValue.integerValue < valueInteger.integerValue);
            } else if (_conditionOperator == SwrveTriggerOperatorNUMBER_EQUALS) {
                return (payloadValue && payloadValue.integerValue == valueInteger.integerValue);
            } else if (_conditionOperator == SwrveTriggerOperatorNUMBER_BETWEEN) {
                NSNumber *lower = nil;
                NSNumber *upper = nil;
                if([valueDictionary objectForKey:@"lower"] != [NSNull null]) {
                    lower = [valueDictionary objectForKey:@"lower"];
                }
                if([valueDictionary objectForKey:@"upper"] != [NSNull null]) {
                    upper = [valueDictionary objectForKey:@"upper"];
                }
                return (valueDictionary && payloadValue.integerValue > lower.integerValue && payloadValue.integerValue < upper.integerValue);
            } else if (_conditionOperator == SwrveTriggerOperatorNUMBER_NOT_BETWEEN) {
                NSNumber *lower = nil;
                NSNumber *upper = nil;
                if([valueDictionary objectForKey:@"lower"] != [NSNull null]) {
                    lower = [valueDictionary objectForKey:@"lower"];
                }
                if([valueDictionary objectForKey:@"upper"] != [NSNull null]) {
                    upper = [valueDictionary objectForKey:@"upper"];
                }
                return (valueDictionary && (payloadValue.integerValue < lower.integerValue || payloadValue.integerValue > upper.integerValue));
            } else {
                return NO;
            }
        } else {
            return NO;
        }
    } else {
        return NO;
    }
}


@end

File: SwrveSDKSwift/SwrveTriggerCondition.swift
import Foundation

@objc public enum SwrveTriggerOperator: UInt {
    case and
    case or
    case equals
    case contains
    case numberGT
    case numberLT
    case numberEquals
    case numberBetween
    case numberNotBetween
    case other
}

@objc public class SwrveTriggerCondition: NSObject {

    // MARK: - Properties

    /// The logical operator of the trigger
    @objc public var triggerOperator: SwrveTriggerOperator

    /// The condition operator for the trigger
    @objc public var conditionOperator: SwrveTriggerOperator

    /// The key used in the condition
    @objc public var key: String

    /// The value to be compared in the condition
    @objc public var value: Any

    /// Initializes the trigger condition with a dictionary and operator
    /// - Parameters:
    ///   - dictionary: A dictionary containing the condition data
    ///   - operatorKey: The operator used for the condition

    @objc public init?(dictionary: [String: Any], operatorKey: String?) {
        guard let key = dictionary["key"] as? String,
            let value = dictionary["value"]
        else {
            return nil
        }

        self.key = key
        self.value = value
        self.triggerOperator = SwrveTriggerCondition.determineSwrveOperator(operatorKey)
        self.conditionOperator = SwrveTriggerCondition.determineSwrveOperator(dictionary["op"] as? String)
        super.init()

        if key.isEmpty || self.conditionOperator == .other {
            return nil
        }
    }

    // MARK: - Methods

    /// Determines the Swrve operator based on a string
    /// - Parameter op: The operator string
    /// - Returns: The corresponding `SwrveTriggerOperator`

    private static func determineSwrveOperator(_ op: String?) -> SwrveTriggerOperator {
        switch op {
        case "and":
            return .and
        case "eq":
            return .equals
        case "or":
            return .or
        case "contains":
            return .contains
        case "number_eq":
            return .numberEquals
        case "number_gt":
            return .numberGT
        case "number_lt":
            return .numberLT
        case "number_between":
            return .numberBetween
        case "number_not_between":
            return .numberNotBetween
        default:
            return .other
        }
    }

    /// Checks if the condition has been fulfilled based on the provided payload
    /// - Parameter payload: A dictionary containing the event data
    /// - Returns: Boolean indicating if the condition is fulfilled

    func hasFulfilledCondition(with payload: [AnyHashable: Any]) -> Bool {
        guard let payloadObject = payload[key], !(payloadObject is NSNull) else {
            return false
        }

        let payloadValue: String
        if let strValue = payloadObject as? String {
            payloadValue = strValue
        } else if let numValue = payloadObject as? NSNumber {
            payloadValue = numValue.stringValue
        } else {
            return false
        }

        let valueString = value as? String
        let valueNumber = value as? NSNumber
        let valueDict = value as? [String: Any]

        switch conditionOperator {
        case .contains:
            return payloadValue.localizedCaseInsensitiveContains(valueString ?? "")
        case .equals:
            return payloadValue.caseInsensitiveCompare(valueString ?? "") == .orderedSame
        case .numberGT:
            return (payloadValue as NSString).integerValue > (valueNumber?.intValue ?? 0)
        case .numberLT:
            return (payloadValue as NSString).integerValue < (valueNumber?.intValue ?? 0)
        case .numberEquals:
            return (payloadValue as NSString).integerValue == (valueNumber?.intValue ?? 0)
        case .numberBetween:
            guard let lower = valueDict?["lower"] as? NSNumber, let upper = valueDict?["upper"] as? NSNumber else {
                return false
            }
            let payloadIntValue = (payloadValue as NSString).integerValue
            return payloadIntValue > lower.intValue && payloadIntValue < upper.intValue
        case .numberNotBetween:
            guard let lower = valueDict?["lower"] as? NSNumber, let upper = valueDict?["upper"] as? NSNumber else {
                return false
            }
            let payloadIntValue = (payloadValue as NSString).integerValue
            return payloadIntValue < lower.intValue || payloadIntValue > upper.intValue
        default:
            return false
        }
    }
}
